client/client.go

```diff
@@ -34,6 +34,50 @@ import (
 	"github.com/canonical/pebble/internals/wsutil"
 )
 
+type Requester interface {
+	// Do performs the HTTP transaction using the provided options.
+	Do(ctx context.Context, opts *RequestOptions) (*RequestResponse, error)
+
+	// Transport returns the HTTP transport in use by the underlying HTTP client.
+	Transport() *http.Transport
+}
+
+type RequestType int
+
+const (
+	RawRequest RequestType = iota
+	SyncRequest
+	AsyncRequest
+)
+
+type RequestOptions struct {
+	Type    RequestType
+	Method  string
+	Path    string
+	Query   url.Values
+	Headers map[string]string
+	Body    io.Reader
+}
+
+type RequestResponse struct {
+	StatusCode int
+	ChangeID   string
+	Result     []byte
+
+	// Only set for RawRequest.
+	Body io.ReadCloser
+}
+
+// DecodeResult decodes the endpoint-specific result payload that is included as part of
+// sync and async request responses. The decoding is performed with the standard JSON
+// package, so the usual field tags should be used to prepare the type for decoding.
+func (resp *RequestResponse) DecodeResult(result interface{}) error {
+	if err := decodeWithNumber(bytes.NewReader(resp.Result), result); err != nil {
```

Comment:
Can we move the `decodeWithNumber` implementation here and get rid of it? It seems to only be used once, and it's only usage is here.

Reply:
I collapsed ```decodeWithNumber()``` into this. It makes sense for me :) thanks for the feedback.